RandLAPACK/gpu_functions/rl_cuda_kernels.cuh

```diff
@@ -0,0 +1,934 @@
+    int64_t* end = idx_mut + k;
+    for (int64_t i = 1; i <= k; ++i, ++curr) {
+        // swap rows IFF mismatched
+        if (int64_t const j = *curr; i != j) {
+            if (threadIdx.x == 0) {
+                std::iter_swap(J + i, J + j);
+                std::iter_swap(curr, std::find(curr, end, i));
+            }
+            __syncthreads();
+        }
+    }
+}
+
+template <typename T>
+__global__ void vec_ell_swap_gpu(
+    int64_t m, 
+    int64_t k,
+    int64_t* J, 
+    int64_t* J_cpy, 
+    int64_t const* idx
+) {
+    for (int64_t i = threadIdx.x + blockDim.x * blockIdx.x; i < k; i += blockDim.x * gridDim.x) {
+        int64_t j = idx[i] - 1;
+        J[i] = J_cpy[j];
+    }
+}
+
+// Transposes the input matrix, copying the transposed version into the buffer.
+// If an option is passes, stores only the upper-triangular portion of the transposed factor.
+// This functioin would require a copy with an adjustible stride.
+template <typename T>
+__global__ void transposition_gpu(
+    int64_t m,
+    int64_t n,
+    const T* A,
+    int64_t lda,
+    T* AT,
+    int64_t ldat,
+    bool copy_upper_triangle
+) {
+    if (copy_upper_triangle) {
+        // Only transposing the upper-triangular portion of the original
+        #pragma unroll
+        for(int64_t i = 0; i < n; ++i){
+            copy_gpu_device(i + 1, &A[i * lda], 1, &AT[i], ldat);
+        }
+    } else {
+        #pragma unroll
+        for(int64_t i = 0; i < n; ++i){
+            copy_gpu_device(m, &A[i * lda], 1, &AT[i], ldat);
+        }
+    }
+}
+
+template <typename T>
+__global__ void copy_mat_gpu(
+    int64_t m,
+    int64_t n,
+    const T* A,
+    int64_t lda,
+    T* A_cpy,
+    int64_t ldac,
+    bool copy_upper_triangle
+) {
+    if (copy_upper_triangle) {
+        // Only copying the upper-triangular portion of the original
+        for (int64_t i = threadIdx.x + blockDim.x * blockIdx.x; i < n; i += blockDim.x * gridDim.x) {
+            for (int64_t j = threadIdx.y + blockDim.y * blockIdx.y; j <= i; j += blockDim.y * gridDim.y) {
+                A_cpy[(ldac * i) + j] = A[(lda * i) + j];
+            }
+        }
+    } else {
+        for (int64_t i = threadIdx.x + blockDim.x * blockIdx.x; i < n; i += blockDim.x * gridDim.x) {
+            for (int64_t j = threadIdx.y + blockDim.y * blockIdx.y; j < m; j += blockDim.y * gridDim.y) {
+                A_cpy[(ldac * i) + j] = A[(lda * i) + j];
+            }
+        }
+    }
+}
+#endif
+
+template <typename T>
+void naive_rank_est(
+    cudaStream_t stream, 
+    int64_t n,
+    const T alpha,
+    const T* A,
+    int64_t lda,
+    int64_t& rank
+) {
+#ifdef USE_CUDA
+    int64_t* rank_device;
+    cudaMalloc(&rank_device, sizeof(int64_t));
+
+    constexpr int threadsPerBlock{128};
+    int64_t num_blocks_ger{(n + threadsPerBlock - 1) / threadsPerBlock};
+    naive_rank_est<<<num_blocks_ger, threadsPerBlock, sizeof(int64_t), stream>>>(n, alpha, A, lda, rank_device);
+#endif
```

Comment:
It seems dangerous to me to have this ifdef here. Will this function be correct if the ifdef is false?

Reply:
If not, then please have an #else which contains a single line, and that line throws an error.

Reply:
Fixed.